File: src/lib/db/knownHosts.ts
/**
 * @file 已知 SSH 主机指纹
 * @author Charlie
 * @description known_hosts 表 CRUD；与后端 SSH 主机密钥校验共用。
 */

import { getDb } from "@/lib/db/client";

export type KnownHostRow = {
  id: number;
  host: string;
  port: number;
  fingerprint: string;
};

export async function listKnownHosts(): Promise<KnownHostRow[]> {
  const db = await getDb();
  return db.select<KnownHostRow[]>(
    "SELECT id, host, port, fingerprint FROM known_hosts ORDER BY host, port",
  );
}

export async function addKnownHost(
  host: string,
  port: number,
  fingerprint: string,
): Promise<number> {
  const db = await getDb();
  const result = await db.execute(
    `INSERT INTO known_hosts (host, port, fingerprint) VALUES ($1, $2, $3)
     ON CONFLICT(host, port) DO UPDATE SET fingerprint = excluded.fingerprint`,
    [host.trim(), port, fingerprint.trim()],
  );
  return result.lastInsertId as number;
}

export async function removeKnownHost(id: number): Promise<void> {
  const db = await getDb();
  await db.execute("DELETE FROM known_hosts WHERE id = $1", [id]);
}

export async function getKnownHostFingerprint(
  host: string,
  port: number,
): Promise<string | null> {
  const db = await getDb();
  const rows = await db.select<{ fingerprint: string }[]>(
    "SELECT fingerprint FROM known_hosts WHERE host = $1 AND port = $2 LIMIT 1",
    [host.trim(), port],
  );
  return rows[0]?.fingerprint ?? null;
}

File: src-tauri/src/services/ssh/known_hosts.rs
//! known_hosts 表读写。
//!
//! Author: Charlie

use anyhow::{Context, Result};
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};

use crate::data_dir;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnownHostRow {
    pub id: i64,
    pub host: String,
    pub port: u16,
    pub fingerprint: String,
}

fn open_db() -> Result<Connection> {
    let path = data_dir::db_path()?;
    Connection::open(path).context("open sqlite db")
}

pub fn lookup_fingerprint(host: &str, port: u16) -> Result<Option<String>> {
    let conn = open_db()?;
    let mut stmt =
        conn.prepare("SELECT fingerprint FROM known_hosts WHERE host = ?1 AND port = ?2 LIMIT 1")?;
    let mut rows = stmt.query(params![host, i64::from(port)])?;
    if let Some(row) = rows.next()? {
        return Ok(Some(row.get(0)?));
    }
    Ok(None)
}

pub fn list_known_hosts() -> Result<Vec<KnownHostRow>> {
    let conn = open_db()?;
    let mut stmt =
        conn.prepare("SELECT id, host, port, fingerprint FROM known_hosts ORDER BY host, port")?;
    let rows = stmt.query_map([], |row| {
        Ok(KnownHostRow {
            id: row.get(0)?,
            host: row.get(1)?,
            port: row.get::<_, i64>(2)? as u16,
            fingerprint: row.get(3)?,
        })
    })?;
    rows.collect::<Result<Vec<_>, _>>()
        .context("list known_hosts")
}

pub fn add_known_host(host: &str, port: u16, fingerprint: &str) -> Result<i64> {
    let conn = open_db()?;
    conn.execute(
        "INSERT INTO known_hosts (host, port, fingerprint) VALUES (?1, ?2, ?3)
         ON CONFLICT(host, port) DO UPDATE SET fingerprint = excluded.fingerprint",
        params![host.trim(), i64::from(port), fingerprint.trim()],
    )
    .context("upsert known_hosts")?;
    Ok(conn.last_insert_rowid())
}

pub fn remove_known_host(id: i64) -> Result<()> {
    let conn = open_db()?;
    conn.execute("DELETE FROM known_hosts WHERE id = ?1", params![id])
        .context("delete known_hosts")?;
    Ok(())
}
